windup/windup-operator
# Rewrite operator - use quarkus-operator-sdk

## How to test this PR using Minikube
1. Init Minikube:
```
minikube start --cpus=8 --memory=10g
minikube addons enable ingress
```

> If your internet speed is slow it might be worth pulling the images manually to speed up your tests. Use the following commands:
```
minikube ssh docker pull quay.io/windupeng/windup-web-openshift:latest
minikube ssh docker pull quay.io/windupeng/windup-web-openshift-messaging-executor:latest
minikube ssh docker pull registry.access.redhat.com/rhscl/postgresql-10-rhel7:1
```

2. Make sure you are using JDK 17 and then:

```
mvn clean test -Dtest=WindupReconcilerTest
```

- The test will detect that you have a k8s client active. The client is automatically enabled when you initialize Minikube.
- You can monitor the resources being created using the Minikube dashboard or the kubectl command.

## How to test this PR using OCP

### Method1: using same approach as in Minikube
- You need to have an OCP instance running. You must connect your terminal to OCP using:
```
oc login ....
```
- You can not deploy the operator in the `default` namespace, therefore you need to create a new project using
```
oc new-project my-project
```

- If you create a project using the web console, make sure your k8s client points to that project using:
```
oc project my-project
```

- Finally just execute the command
```
mvn clean test -Dtest=WindupReconcilerTest
```

### Method2: creating a catalog
> NOTE: for each image pushed to quay.io make sure you don't have them as private but public so OCP and OPM can pull your images.

- Install OPM (only if you have not done it yet)

```
export OPERATOR_SDK_DL_URL=https://github.com/operator-framework/operator-registry/releases/download/v1.26.2
curl -LO ${OPERATOR_SDK_DL_URL}/${OS}-${ARCH}-opm
chmod +x ${OS}-${ARCH}-opm && sudo mv ${OS}-${ARCH}-opm /usr/local/bin/opm
```

- Create a container image for the operator

```shell
mvn clean package -DskipTests \
-Dquarkus.native.container-build=true \
-Dquarkus.container-image.build=true \
-Dquarkus.container-image.push=false \
-Dquarkus.container-image.registry=quay.io \
-Dquarkus.container-image.group=$USER \
-Dquarkus.container-image.name=windup-operator \
-Dquarkus.operator-sdk.bundle.package-name=windup-operator \
-Dquarkus.operator-sdk.bundle.channels=alpha \
-Dquarkus.application.version=test \
-P native
podman push quay.io/$USER/windup-operator:test
```

- Enrich bundle with cluster permissions

```shell
groovy scripts/enrichCSV.groovy target/bundle/windup-operator/manifests/windup-operator.clusterserviceversion.yaml
```

- Create the bundle:

```shell
BUNDLE_IMAGE=quay.io/$USER/windup-operator-bundle:test
podman build -t $BUNDLE_IMAGE -f target/bundle/windup-operator/bundle.Dockerfile target/bundle/windup-operator
podman push $BUNDLE_IMAGE
```

- Create catalog image:
```shell
CATALOG_IMAGE=quay.io/$USER/windup-operator-catalog:test
opm index add \
    --bundles $BUNDLE_IMAGE \
    --tag $CATALOG_IMAGE \
    --build-tool podman
podman push $CATALOG_IMAGE
```

- Instantiate the catalog:

```shell
cat <<EOF | kubectl apply -f -
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: windup-catalog-source
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: $CATALOG_IMAGE
EOF
```

At this point you can see the Operator in the marketplace of OCP ready for you to test it using the web console of OCP
